Maplike and Setlike objects in Map and Set (#27160)
mdn/content: files/en-us/web/javascript/reference/global_objects/set/index.md

The [`has`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/has) method checks if a value is in the set, using an approach that is, on average, quicker than testing most of the elements that have previously been added to the set. In particular, it is, on average, faster than the [`Array.prototype.includes`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes) method when an array has a `length` equal to a set's `size`.

> ### Set-like objects

Reviewer: Okay, this is a super weird issue: JavaScript is going to soon have its own definition of "set-like objects": [set-methods](https://github.com/tc39/proposal-set-methods) accept "Set-like objects" as arguments, which are objects with `has`, `keys`, and `size`. Maybe we should say specifically: "Set-like objects in Web IDL" or "Set-like browser APIs"?

Author: Makes sense.

Browser **`Set`-like objects** (or "setlike objects") are [Web API](/en-US/docs/Web/API) interfaces that behave in many ways like a [`Set`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set).

Just like `Set`, elements can be iterated in the same order that they were added to the object.

> `Set`-like objects also have properties and methods that share the same name and behavior.

Reviewer: ```suggestion
`Set`-like objects also have properties and methods that share the same name and behavior with `Set`.
```

> Just like `Set` they are ordered, and entries are iterated in the same sequence that they were added to the object, using all of the same looping constructs.
> `Set`-like objects also share some of the same properties and methods.

Reviewer: ```suggestion
Just like `Set`, they are ordered, and elements can be iterated in the order that they were added to the object.
`Set`-like objects also share with `Set` some properties and methods of the same name.
```

Author: ```suggestion
Just like `Set`, elements can be iterated in the same order that they were added to the object.
`Set`-like objects also have properties and methods that share the same name and behavior.
```

However unlike `Set` they only allow a specific predefined type for each entry.

The allowed types are set in the specification IDL definition.
For example, {{domxref("GPUSupportedFeatures")}} is a `Set`-like object that must use strings as the key/value.
This is defined in the specification IDL below:

```webidl
interface GPUSupportedFeatures {
    readonly setlike<DOMString>;

> };

Reviewer: Do we use 2 or 4 spaces to indent web IDL?

Author: I don't know. I just ran prettier.

>     readonly setlike<DOMString>;

Reviewer: ```suggestion
  readonly setlike<DOMString>;
```

```

`Set`-like objects are either read-only or read-writable (see the `readonly` keyword in the IDL above).

> - Read-only `Set`-like objects have the property [`size`](#map.prototype.size), and the methods: [`entries()`](#map.prototype.entries), [`forEach()`](#map.prototype.foreach), [`has()`](#map.prototype.has), [`keys()`](#map.prototype.keys), [`values()`](#map.prototype.values), and [`@@iterator`](#map.prototypeiterator).
> - Writeable `Map`-like objects additionally have the methods: [`clear()`](#map.prototype.clear), [`delete()`](#map.prototype.delete), and [`add()`](#map.prototype.add).

Reviewer: ```suggestion
- Read-only `Set`-like objects have the property [`size`](#set.prototype.size), and the methods: [`entries()`](#set.prototype.entries), [`forEach()`](#set.prototype.foreach), [`has()`](#set.prototype.has), [`keys()`](#set.prototype.keys), [`values()`](#set.prototype.values), and [`@@iterator`](#set.prototypeiterator).
- Writeable `Set`-like objects additionally have the methods: [`clear()`](#set.prototype.clear), [`delete()`](#set.prototype.delete), and [`add()`](#set.prototype.add).
```

Author: Good catch.

> The methods and properties have the same behaviour as the equivalent entities in `Set`, except for the restriction on the types of the entry.

Reviewer: ```suggestion
The methods and properties have the same behavior as the equivalent entities in `Set`, except for the restriction on the types of the entry.
```

> The following interfaces are read-only `Set`-like objects:

Reviewer: ```suggestion
The following are examples of read-only `Set`-like browser objects:
```

- {{domxref("GPUSupportedFeatures")}}
- {{domxref("XRAnchorSet")}}

> The following interfaces are writable `Set`-like objects:

Reviewer: ```suggestion
The following are examples of writable `Set`-like browser objects:
```
